# Simulation using Python, Lab 2, Part A: Monte Carlo Integration

In [ ]:
import math
import numpy as np
import random 
import matplotlib.pyplot as plt
from scipy import stats

**Example**
Compute $\int_0^\pi sin(x) dx$ using MCS

In [ ]:
#Let's define the integral
def integral1(x):
  return (math.sin(x))

Solve above example for N = 10, 100, 1000, 10000 and observe the results!

In [ ]:
N = 10**4 ## number of random points
v=[10,100,1000,10000]
for j in v:
  lb = 0.0  #lower bound
  ub = math.pi	#upper bound

  eval_data = [] 	#Defining eval_data as a list data type

  for i in range(N):
    v = random.uniform(lb, ub)
    eval_data.append(integral1(v))

  val = (ub-lb)*np.mean(eval_data)

  print('Integral Estimate if N=',j,'  : ', val)

Integral Estimate if N= 10   :  2.012316962399818
Integral Estimate if N= 100   :  1.9902484529713984
Integral Estimate if N= 1000   :  2.022785497998896
Integral Estimate if N= 10000   :  2.0168261358718036


Integral Estimate if N= 10   :  2.006486836990008

Integral Estimate if N= 100   :  1.9754021108222473

Integral Estimate if N= 1000   :  2.012866381558127

Integral Estimate if N= 10000   :  1.9925289576982554

as we increase N we get more data point that increase accuracy of integral as we can see from above it's completely depend on how much data point is nearer to our known solution if we take less sample then also possibility of getting known result if random data point is nearer to our known solution but chance is less compare to when we take large no of data point

### Building Confidence Interval
Let's run the model M=10 times, each with N=100 data points. Use the M data points to build the 95% CI of the mean estimate of the integral

In [ ]:
 #Number of batches or sample sets. as M
N = 100 ## number of random points per batch
M=10
lb = 0.0  #lower bound
ub = math.pi	#upper bound

val=[]
for j in range(M): #Outer-loop for sample sets
    eval_data = []

    for i in range(N): #Inner-loop for function evaluation
      v = random.uniform(lb, ub)
      eval_data.append(integral1(v))
    
    val.append((ub-lb)*np.mean(eval_data))

total_mean = np.mean(val)  #mean
print("Integral Mean =",total_mean)
sd = np.std(val)	#standard deviation
print('Standard deviation = ',sd)

alpha=0.05 #since we want to construct 95% CI
dof = M -1	# Degrees of freedom
t_value = stats.t.ppf((1-alpha/2),dof)
print("t-value from student t-dist with dof ", M-1, " and alpha ", alpha, " is ", t_value)
t_hw = t_value*sd/math.sqrt(M)		#Halfwidth
print("Halfwidth = ",t_hw)
print("95% CI of Integral is [", total_mean - t_hw, ", ", total_mean + t_hw, "]" )

Integral Mean = 1.978203325891549
Standard deviation =  0.05484454935888371
t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915
Halfwidth =  0.03923342713773418
95% CI of Integral is [ 1.9389698987538149 ,  2.0174367530292834 ]


Integral Mean = 1.978203325891549

Standard deviation =  0.05484454935888371

t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

Halfwidth =  0.03923342713773418

95% CI of Integral is [ 1.9389698987538149 ,  2.0174367530292834 ]

In [ ]:
 #Number of batches or sample sets. as M
N = 100 ## number of random points per batch
lb = 0.0  #lower bound
ub = math.pi	#upper bound
for M in [10,100,1000,10000]:
  val=[]
  for j in range(M): #Outer-loop for sample sets
    eval_data = []

    for i in range(N): #Inner-loop for function evaluation
      v = random.uniform(lb, ub)
      eval_data.append(integral1(v))
    
    val.append((ub-lb)*np.mean(eval_data))
  print('\n \n \n when M=',M)
  total_mean = np.mean(val)  #mean
  print("Integral Mean =",total_mean)
  sd = np.std(val)	#standard deviation
  print('Standard deviation = ',sd)

  alpha=0.05 #since we want to construct 95% CI
  dof = M -1	# Degrees of freedom
  t_value = stats.t.ppf((1-alpha/2),dof)
  print("t-value from student t-dist with dof ", M-1, " and alpha ", alpha, " is ", t_value)
  t_hw = t_value*sd/math.sqrt(M)		#Halfwidth
  print("Halfwidth = ",t_hw)
  print("95% CI of Integral is [", total_mean - t_hw, ", ", total_mean + t_hw, "]" )


 
 
 when M= 10
Integral Mean = 1.986086888831189
Standard deviation =  0.09753653121397261
t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915
Halfwidth =  0.06977343118657392
95% CI of Integral is [ 1.916313457644615 ,  2.055860320017763 ]

 
 
 when M= 100
Integral Mean = 1.9744462561079825
Standard deviation =  0.0999513396612537
t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827
Halfwidth =  0.019832514248186172
95% CI of Integral is [ 1.9546137418597964 ,  1.9942787703561686 ]

 
 
 when M= 1000
Integral Mean = 1.9923211118935464
Standard deviation =  0.09514434992762986
t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487
Halfwidth =  0.005904152725337722
95% CI of Integral is [ 1.9864169591682088 ,  1.998225264618884 ]

 
 
 when M= 10000
Integral Mean = 2.000192778931527
Standard deviation =  0.09727160241709652
t-value from student t-dist with dof  9999  and alpha  0.05  is  1.960201263

#To Do

1. Estimate $I = \int_0^{9/10} \int_0^1 \int_0^{11/10} (4-x^2-y^2-z^2)dz dy dx$ using MCS.  Bound the expected value with a 95% CI

In [ ]:
def integral3(x,y,z):
  return 4-x**2-y**2-z**2

In [ ]:
for M in [10,100,1000,10000]:
  val=[]
  for i in range(M):
    eval_data1=[]
    for j in range(100):
      x=random.uniform(0,9/10)
      y=random.uniform(0,1)
      z=random.uniform(0,11/10)
      eval_data1.append(integral3(x,y,z))
    val.append((9/10)*(11/10)*np.mean(eval_data1))
  print("\n \n when M=",M)  
  print("\n integral mean=",np.mean(val))
  print("\n standard deviation=",np.std(val))
  alpha=0.05 #since we want to construct 95% CI
  dof = M -1	# Degrees of freedom
  t_value = stats.t.ppf((1-alpha/2),dof)
  print("\n t-value from student t-dist with dof ", M-1, " and alpha ", alpha, " is ", t_value)
  t_hw = t_value*np.std(val)/math.sqrt(M)		#Halfwidth
  print("\n Halfwidth = ",t_hw)
  print("\n 95% CI of Integral is [", np.mean(val) - t_hw, ", ",np.mean(val) + t_hw, "]" )



 
 when M= 10

 integral mean= 2.9608496897350784

 standard deviation= 0.03967426342449903

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.028381258329296926

 95% CI of Integral is [ 2.9324684314057814 ,  2.9892309480643755 ]

 
 when M= 100

 integral mean= 2.959822924297788

 standard deviation= 0.04893274321639435

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.009709317857379118

 95% CI of Integral is [ 2.950113606440409 ,  2.969532242155167 ]

 
 when M= 1000

 integral mean= 2.962427845755261

 standard deviation= 0.05068822303445218

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.0031454417461365095

 95% CI of Integral is [ 2.9592824040091243 ,  2.9655732875013974 ]

 
 when M= 10000

 integral mean= 2.964405450115872

 standard deviation= 0.052160250268186534

 t-value from student t-dist with dof  9999  and alpha  0


 
 when M= 10

 integral mean= 2.9608496897350784

 standard deviation= 0.03967426342449903

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.028381258329296926

 95% CI of Integral is [ 2.9324684314057814 ,  2.9892309480643755 ]

 
 when M= 100

 integral mean= 2.959822924297788

 standard deviation= 0.04893274321639435

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.009709317857379118

 95% CI of Integral is [ 2.950113606440409 ,  2.969532242155167 ]

 
 when M= 1000

 integral mean= 2.962427845755261

 standard deviation= 0.05068822303445218

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.0031454417461365095

 95% CI of Integral is [ 2.9592824040091243 ,  2.9655732875013974 ]

 
 when M= 10000

 integral mean= 2.964405450115872

 standard deviation= 0.052160250268186534

 t-value from student t-dist with dof  9999  and alpha  0.05  is  1.9602012636213575

 Halfwidth =  0.001022445884865055

 95% CI of Integral is [ 2.963383004231007 ,  2.9654278960007368 ]

2. Estimate $I = \int_0^\inf x^{\alpha -1} e^{-x} dx$, with $\alpha$ = 1.9 using MCS.  Option 1: Try with $x$ ~ Expo(1).  Option 2: Try with $x$ ~ Erlang(k=2, $\beta$=1). Compare the results across options for same value of N=100 and N=1000 

In [ ]:
def integral4(x):
  return (x**0.9)*math.exp(-x)

In [ ]:
N=100
for M in [10,100,1000,10000]:
  val=[]
  for i in range(M):
    eval_data1=[]
    for j in range(N):
      x=np.random.exponential(1)
      eval_data1.append((1/(math.exp(-x)))*integral4(x))
    val.append(np.mean(eval_data1))
  print("\n \n when M=",M)  
  print("\n integral mean=",np.mean(val))
  print("\n standard deviation=",np.std(val))
  alpha=0.05 #since we want to construct 95% CI
  dof = M -1	# Degrees of freedom
  t_value = stats.t.ppf((1-alpha/2),dof)
  print("\n t-value from student t-dist with dof ", M-1, " and alpha ", alpha, " is ", t_value)
  t_hw = t_value*np.std(val)/math.sqrt(M)		#Halfwidth
  print("\n Halfwidth = ",t_hw)
  print("\n 95% CI of Integral is [", np.mean(val) - t_hw, ", ",np.mean(val) + t_hw, "]" )



 
 when M= 10

 integral mean= 0.9444169656181798

 standard deviation= 0.08628718923602709

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.061726136715228426

 95% CI of Integral is [ 0.8826908289029514 ,  1.0061431023334082 ]

 
 when M= 100

 integral mean= 0.9686875415002841

 standard deviation= 0.08555941898578358

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.01697684495128256

 95% CI of Integral is [ 0.9517106965490015 ,  0.9856643864515667 ]

 
 when M= 1000

 integral mean= 0.9664116220590375

 standard deviation= 0.0873632525273834

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.005421299172218408

 95% CI of Integral is [ 0.9609903228868191 ,  0.971832921231256 ]

 
 when M= 10000

 integral mean= 0.9607695835908003

 standard deviation= 0.08613555308938417

 t-value from student t-dist with dof  9999  and alpha  0

when N=100
 
 when M= 10

 integral mean= 1.0094300074022597

 standard deviation= 0.09806207825970148

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.07014938489513904

 95% CI of Integral is [ 0.9392806225071207 ,  1.0795793922973989 ]

 
 when M= 100

 integral mean= 0.9584145927411224

 standard deviation= 0.09118465394288212

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.018093013607091975

 95% CI of Integral is [ 0.9403215791340305 ,  0.9765076063482143 ]

 
 when M= 1000

 integral mean= 0.9596549698881534

 standard deviation= 0.08304374758739078

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.005153253651039634

 95% CI of Integral is [ 0.9545017162371138 ,  0.9648082235391929 ]

 
 when M= 10000

 integral mean= 0.9628577341316904

 standard deviation= 0.08631572901159436

 t-value from student t-dist with dof  9999  and alpha  0.05  is  1.9602012636213575

 Halfwidth =  0.0016919620107892594

 95% CI of Integral is [ 0.9611657721209012 ,  0.9645496961424797 ]

In [ ]:
N=1000
for M in [10,100,1000,10000]:
  val=[]
  for i in range(M):
    eval_data1=[]
    for j in range(N):
      x=np.random.exponential(1)
      eval_data1.append((1/(math.exp(-x)))*integral4(x))
    val.append(np.mean(eval_data1))
  print("\n \n when M=",M)  
  print("\n integral mean=",np.mean(val))
  print("\n standard deviation=",np.std(val))
  alpha=0.05 #since we want to construct 95% CI
  dof = M -1	# Degrees of freedom
  t_value = stats.t.ppf((1-alpha/2),dof)
  print("\n t-value from student t-dist with dof ", M-1, " and alpha ", alpha, " is ", t_value)
  t_hw = t_value*np.std(val)/math.sqrt(M)		#Halfwidth
  print("\n Halfwidth = ",t_hw)
  print("\n 95% CI of Integral is [", np.mean(val) - t_hw, ", ",np.mean(val) + t_hw, "]" )



 
 when M= 10

 integral mean= 0.9556795988249185

 standard deviation= 0.015714653893423965

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.01124158618731531

 95% CI of Integral is [ 0.9444380126376032 ,  0.9669211850122338 ]

 
 when M= 100

 integral mean= 0.9600077837784198

 standard deviation= 0.02341493560107212

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.004646031213813144

 95% CI of Integral is [ 0.9553617525646066 ,  0.9646538149922329 ]

 
 when M= 1000

 integral mean= 0.9616198677205452

 standard deviation= 0.027330698776548083

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.0016959979209451776

 95% CI of Integral is [ 0.9599238697996 ,  0.9633158656414904 ]

 
 when M= 10000

 integral mean= 0.9618736912478049

 standard deviation= 0.027454468087297536

 t-value from student t-dist with dof  9999  and alpha

when N=1000
 
 when M= 10

 integral mean= 0.9556795988249185

 standard deviation= 0.015714653893423965

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.01124158618731531

 95% CI of Integral is [ 0.9444380126376032 ,  0.9669211850122338 ]

 
 when M= 100

 integral mean= 0.9600077837784198

 standard deviation= 0.02341493560107212

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.004646031213813144

 95% CI of Integral is [ 0.9553617525646066 ,  0.9646538149922329 ]

 
 when M= 1000

 integral mean= 0.9616198677205452

 standard deviation= 0.027330698776548083

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.0016959979209451776

 95% CI of Integral is [ 0.9599238697996 ,  0.9633158656414904 ]

 
 when M= 10000

 integral mean= 0.9618736912478049

 standard deviation= 0.027454468087297536

 t-value from student t-dist with dof  9999  and alpha  0.05  is  1.9602012636213575

 Halfwidth =  0.0005381628303677286

 95% CI of Integral is [ 0.9613355284174371 ,  0.9624118540781726 ]

 as we increase N we get more data point that increase accuracy of integral as we can see from above it's completely depend on how much data point is nearer to our known solution if we take less sample then also possibility of getting known result if random data point is nearer to our known solution but chance is less compare to when we take large no of data point

In [ ]:
N=100
for M in [10,100,1000]:
  val=[]
  for i in range(M):
    eval_data1=[]
    for j in range(N):
      x=np.random.gamma(2,1/1)
      eval_data1.append((1/(stats.erlang.pdf(x,2)))*integral4(x))
    val.append(np.mean(eval_data1))
  print("\n \n when M=",M)  
  print("\n integral mean=",np.mean(val))
  print("\n standard deviation=",np.std(val))
  alpha=0.05 #since we want to construct 95% CI
  dof = M -1	# Degrees of freedom
  t_value = stats.t.ppf((1-alpha/2),dof)
  print("\n t-value from student t-dist with dof ", M-1, " and alpha ", alpha, " is ", t_value)
  t_hw = t_value*np.std(val)/math.sqrt(M)		#Halfwidth
  print("\n Halfwidth = ",t_hw)
  print("\n 95% CI of Integral is [", np.mean(val) - t_hw, ", ",np.mean(val) + t_hw, "]" )



 
 when M= 10

 integral mean= 0.9601388044932492

 standard deviation= 0.0075955916651132115

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.005433558952434533

 95% CI of Integral is [ 0.9547052455408146 ,  0.9655723634456838 ]

 
 when M= 100

 integral mean= 0.9604799478839098

 standard deviation= 0.008339859861763284

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.0016548091310917567

 95% CI of Integral is [ 0.958825138752818 ,  0.9621347570150015 ]

 
 when M= 1000

 integral mean= 0.96177162472936

 standard deviation= 0.00796528960532783

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.0004942835425032796

 95% CI of Integral is [ 0.9612773411868567 ,  0.9622659082718632 ]


when N=100
 
 when M= 10

 integral mean= 0.9601388044932492

 standard deviation= 0.0075955916651132115

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.005433558952434533

 95% CI of Integral is [ 0.9547052455408146 ,  0.9655723634456838 ]

 
 when M= 100

 integral mean= 0.9604799478839098

 standard deviation= 0.008339859861763284

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.0016548091310917567

 95% CI of Integral is [ 0.958825138752818 ,  0.9621347570150015 ]

 
 when M= 1000

 integral mean= 0.96177162472936

 standard deviation= 0.00796528960532783

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.0004942835425032796

 95% CI of Integral is [ 0.9612773411868567 ,  0.9622659082718632 ]

In [ ]:
N=1000
for M in [10,100,1000]:
  val=[]
  for i in range(M):
    eval_data1=[]
    for j in range(N):
      x=np.random.gamma(2,1/1)
      eval_data1.append((1/((stats.erlang.pdf(x,2))))*integral4(x))
    val.append(np.mean(eval_data1))
  print("\n \n when M=",M)  
  print("\n integral mean=",np.mean(val))
  print("\n standard deviation=",np.std(val))
  alpha=0.05 #since we want to construct 95% CI
  dof = M -1	# Degrees of freedom
  t_value = stats.t.ppf((1-alpha/2),dof)
  print("\n t-value from student t-dist with dof ", M-1, " and alpha ", alpha, " is ", t_value)
  t_hw = t_value*np.std(val)/math.sqrt(M)		#Halfwidth
  print("\n Halfwidth = ",t_hw)
  print("\n 95% CI of Integral is [", np.mean(val) - t_hw, ", ",np.mean(val) + t_hw, "]" )



 
 when M= 10

 integral mean= 0.9619537721696126

 standard deviation= 0.0021163815438699288

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.0015139681530379203

 95% CI of Integral is [ 0.9604398040165747 ,  0.9634677403226506 ]

 
 when M= 100

 integral mean= 0.9619560002478904

 standard deviation= 0.0025602560224060257

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.000508010339986023

 95% CI of Integral is [ 0.9614479899079044 ,  0.9624640105878765 ]

 
 when M= 1000

 integral mean= 0.9618425793141374

 standard deviation= 0.002511489524877667

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.0001558496929585675

 95% CI of Integral is [ 0.9616867296211788 ,  0.961998429007096 ]


N=1000
 
 when M= 10

 integral mean= 0.9619537721696126

 standard deviation= 0.0021163815438699288

 t-value from student t-dist with dof  9  and alpha  0.05  is  2.2621571627409915

 Halfwidth =  0.0015139681530379203

 95% CI of Integral is [ 0.9604398040165747 ,  0.9634677403226506 ]

 
 when M= 100

 integral mean= 0.9619560002478904

 standard deviation= 0.0025602560224060257

 t-value from student t-dist with dof  99  and alpha  0.05  is  1.9842169515086827

 Halfwidth =  0.000508010339986023

 95% CI of Integral is [ 0.9614479899079044 ,  0.9624640105878765 ]

 
 when M= 1000

 integral mean= 0.9618425793141374

 standard deviation= 0.002511489524877667

 t-value from student t-dist with dof  999  and alpha  0.05  is  1.9623414611334487

 Halfwidth =  0.0001558496929585675

 95% CI of Integral is [ 0.9616867296211788 ,  0.961998429007096 ]

 as we increase N we get more data point that increase accuracy of integral as we can see from above it's completely depend on how much data point is nearer to our known solution if we take less sample then also possibility of getting known result if random data point is nearer to our known solution but chance is less compare to when we take large no of data point